src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,155 @@
 package se.liu.ida.hefquin.engine.datastructures.impl;
 
-import java.util.Iterator;
+import java.util.*;
 
 import org.apache.jena.graph.Node;
 import org.apache.jena.sparql.core.Var;
 
+import org.apache.jena.sparql.engine.binding.Binding;
 import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
 
 public class SolutionMappingsHashTable extends SolutionMappingsIndexBase
 {
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Set<Var> joinVariables;
+
+	public SolutionMappingsHashTable(final ExpectedVariables... inputVars){
+		assert inputVars.length == 2;
+		final Set<Var> certainVars = new HashSet<>( inputVars[0].getCertainVariables());
+		certainVars.retainAll( inputVars[1].getCertainVariables() );
+
+		this.joinVariables = certainVars;
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		return map.size();
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		return map.isEmpty();
 	}
 
 	@Override
 	public boolean contains( final Object o ) {
-		// TODO Auto-generated method stub
-		return false;
+		return map.containsKey(o);
 	}
 
 	@Override
 	public Iterator<SolutionMapping> iterator() {
```

Comment:
```suggestion
	public Iterator<SolutionMapping> iterator() {
		// The following implementation of this method is inefficient because,
		// each time it is called, it creates and populates a list of solution
		// mappings by iterating over the content of this index. However,
		// for a thread-safe implementation, we may have to live with this.
		// TODO: is there a more efficient way to implement this method?
```